skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt

```diff
@@ -52,50 +50,70 @@ object Library {
     // This function does the following: on request to load given resource,
     // it checks if resource with given name is found in content-derived directory
     // in Skiko's home, and if not - unpacks it. Also, it could load additional
-    // localization resource on platforms wher it is needed.
+    // localization resource on platforms where it is needed.
     @Synchronized
     fun load() {
         if (loaded.compareAndExchange(false, true)) return
+
+        // Find/unpack a usable copy of the native library.
+        findAndLoad()
+
+        // TODO move properties to SkikoProperties
+        Setup.init()
+
+        try {
+            // Init code executed after library was loaded.
+            org.jetbrains.skia.impl.Library._nAfterLoad()
+        } catch (t: Throwable) {
+            t.printStackTrace()
+        }
+    }
+
+    private fun findAndLoad() {
         val name = "skiko-$hostId"
         val platformName = System.mapLibraryName(name)
         val icu = if (hostOs.isWindows) "icudtl.dat" else null
 
+        // First try: system property is set.
         if (skikoLibraryPath != null) {
             val library = File(File(skikoLibraryPath), platformName)
             loadLibraryOrCopy(library)
-            if (icu != null && copyDir != null) {
+            if (icu != null && copyDir != null)
                 unpackIfNeeded(copyDir!!, icu, true)
-            }
-        } else {
-            val hashResourceStream = Library::class.java.getResourceAsStream(
-                "/$platformName.sha256"
-            ) ?: throw LibraryLoadException(
-                "Cannot find $platformName.sha256, proper native dependency missing."
-            )
-            val hash = hashResourceStream.use {
-                BufferedReader(InputStreamReader(it)).lines().toArray()[0] as String
-            }
-            val cacheDir = File(File(cacheRoot), hash)
-            cacheDir.mkdirs()
-            val library = unpackIfNeeded(cacheDir, platformName, false)
-            loadLibraryOrCopy(library)
-            if (icu != null) {
-                if (copyDir != null) {
-                    unpackIfNeeded(copyDir!!, icu, true)
-                } else {
-                    unpackIfNeeded(cacheDir, icu, false)
-                }
-            }
+            return
         }
 
-        // TODO move properties to SkikoProperties
-        Setup.init()
+        // Second try: load it from the bin/ or lib/ directory relative to the JVM home.
+        // The user might have placed the native files alongside the other JVM libraries
+        // for signing purposes, so we'll find it here if so.
+        val jvmFiles = File(System.getProperty("java.home"), if (hostOs.isWindows) "bin" else "lib")
```

Comment:
Wouldn't it be natural to search in java.library.home instead?

Reply:
Hmm, where do you see that property? I don't seem to have it when I run `java -XshowSettings:properties` and I can't find it by grepping the OpenJDK sources either.

Reply:
Smth like that could print all the props

```java
import java.util.*;

class a {
    public static void main(String[] args) {
        Properties p = System.getProperties();
        Enumeration keys = p.keys();
        while (keys.hasMoreElements()) {
            String key = (String)keys.nextElement();
            String value = (String)p.get(key);
            System.out.println(key + ": " + value);
        }
    }
}
```
and then `java a.java`.

Reply:
Scanning `java.library.path` is what `System.loadLibrary` does, but in our conversation on Friday you expressed that this would not be ideal because it would potentially pick up library conflicts. And indeed as you see, `java.library.path` has quite a few legacy directories in it where there may be system-installed stuff as well as `.`. Checking the lib directory in the JVM install path only, avoids the potential for conflicts.

Reply:
Well, I am just a bit concerned about cases where we search in JVM home  directory, what is desired behavior here? Why we assume that people put Skiko’s native library to the JDK distribution? Per https://javarevisited.blogspot.com/2013/04/what-is-javalibrarypath-how-to-set-in-Eclipse.html?m=1 and similar docs it seems that java.libary.path is the property intended for exactly the usecase we discuss.

Reply:
And we need to scan that explicitly to be able to make a copy for several classloaders case.

Reply:
The code doesn't assume that - if someone does put it alongside the other JNI libs there, it'll work after this change, but if they don't there should be no difference.

I think the code in this PR will happily make copies for multiple classloaders because `loadLibraryOrCopy` is still called.

`java.library.path` can be used. It's equivalent to starting with `loadLibrary` and only continuing on the old codepath if it doesn't work. It means if there's a skiko library in the user's current working directory or , that will override the one in the JAR or JVM. Also consider that `java.library.path` contains many other entries on Windows, for instance, mine has:

```
    java.library.path = C:\Program Files\Amazon Corretto\jdk16.0.2_7\bin
        C:\Windows\Sun\Java\bin
        C:\Windows\system32
        C:\Windows
        C:\Program Files (x86)\Parallels\Parallels Tools\Applications
        C:\Program Files\Amazon Corretto\jdk16.0.2_7\bin
        C:\Windows\system32
        C:\Windows
        C:\Windows\System32\Wbem
        C:\Windows\System32\WindowsPowerShell\v1.0\
        C:\Windows\System32\OpenSSH\
        C:\Program Files\CMake\bin
        C:\Program Files\dotnet\
        C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\
        C:\Users\mike\AppData\Local\Microsoft\WindowsApps
        .
```

i.e. the entire PATH is searched. So that's a lot of places where Skiko's library might be found. In particular, if you have two programs using Skiko and one is on the PATH, then that will potentially take precedence (unless `java.library.path` is overridden from the command line). That's probably not what's wanted.

I think there's really only a handful of places an app might want to load a JNI library from. The JAR itself is one, the source tree during development is another, and the bin/lib directory is the other one because that's where all the JNI libs used by e.g. Swing, JavaFX, etc go by convention.

Reply:
Generally we could merge this PR, if it will make your life easier (and also makes code a bit more readable), although I'm still not 100% convinced that this is indeed "the right thing" ®.

Reply:
Thanks! It's an incremental improvement and we can refine it further if need be.